Dispatch hash operations through the driver wrapper layer

stevew817 opened the pull request:
## Description
This PR is (mostly) a refactoring of existing code, in order to dispatch `psa_hash_xxx` calls through the driver wrapper layer. The hashing implementation on top of Mbed TLS code that used to reside inside of `psa_crypto.c` has been moved out into its own compile unit `psa_crypto_hash.c` as defined by the documentation work from @ronald-cron-arm in #3933 

There is one other change: the implementation of `psa_hash_compare` now first calls the all-in-one `psa_hash_compute` function to calculate the hash into a local buffer, and then does a time-constant memcmp. This way of setting up the call tree will lead to more efficient execution when plugging in actual hardware drivers, since you'll only hit the driver once instead of potentially up to three times (setup/update/verify).

Supersedes #4043

## Status
**READY**

## Requires Backporting
NO  
(new functionality)

## Migrations
NO
(not present in earlier releases)

## Todos
- [X] Tests
- [X] Documentation

## Steps to test or reproduce
Test coverage through existing test cases, since the previous implementation has been fully moved inside of a 'software driver'.

ronald-cron-arm commented:
CI is failing.
Error in all_sh-freebsd-test_clang_opt:
******************************************************************

* test_clang_opt: build/test: clang -O0, full config 

* Thu Feb 18 17:53:26 UTC 2021

******************************************************************

  CC    aes.c

  CC    aesni.c

  CC    arc4.c

  CC    aria.c

  CC    asn1parse.c

  CC    asn1write.c

  CC    base64.c

  CC    bignum.c

  CC    blowfish.c

  CC    camellia.c

  CC    ccm.c

  CC    chacha20.c

  CC    chachapoly.c

  CC    cipher.c

In file included from cipher.c:29:

In file included from ../include/mbedtls/cipher_internal.h:36:

In file included from ../include/psa/crypto.h:3763:

../include/psa/crypto_struct.h:98:43: error: missing field 'ctx' initializer [-Werror,-Wmissing-field-initializers]

    const struct psa_hash_operation_s v = PSA_HASH_OPERATION_INIT;

                                          ^

../include/psa/crypto_struct.h:95:36: note: expanded from macro 'PSA_HASH_OPERATION_INIT'

#define PSA_HASH_OPERATION_INIT {{0}}

                                   ^

1 error generated.

make[1]: *** [Makefile:270: cipher.o] Error 1

make: *** [Makefile:17: lib] Error 2

^^^^test_clang_opt: build/test: clang -O0, full config: command make CC=clang CFLAGS=-O0 -std=c99 -pedantic -Wall -Wextra -Werror -> 2^^^^

(skipped because the build failed)

ronald-cron-arm commented:
Other jobs are failing but it seems it is always the same error as above.

stevew817 commented:
@ronald-cron-arm should be fixed now. Strange that the local test runs are not complaining about that.

mstarzyk-mobica commented:
This PR is changing the API. Please add a changelog entry.

stevew817 commented:
@mstarzyk-mobica It is not changing any public API. The content of the struct is checked by the CI ABI checker, but changing these hasn't required a changelog entry before...

gilles-peskine-arm commented:
I confirm that `psa/crypto_struct.h` is a private header, so changing it is not an API change, only an ABI change. No changelog entry is necessary just because `psa/crypto_struct.h` changes. I don't think this PR should have a changelog entry since the feature it's adding (driver dispatch) is not in its final intended form, but requires modifying a library source file (`library/psa_crypto_driver_wrappers.c`).

mstarzyk-mobica commented:
> @mstarzyk-mobica It is not changing any public API. The content of the struct is checked by the CI ABI checker, but changing these hasn't required a changelog entry before...

Ah, alright! :)

mstarzyk-mobica reviewed (dismissed):
Looks good

ronald-cron-arm reviewed (changes requested):
Please find my comments below. I see four principal issues with the current version:
1) The hash operation context, see comment in crypto_struct.h
2) The addition of checks that pointers are not NULL where I think the Mbed TLS avoid on-purpose to do them (see the Parameter conventions section of the PSA spec for the justification of this).
3) Some code in the wrapper and in psa_crypto_hash.c would be better in psa_crypto.c in my opinion. For code being in psa_crypto.c and not in the wrapper whenever possible the rationale to me is for the wrapper code to do as little as possible to ease its generation.
4) The absence of transparent test driver hash entry points (arguably this could be done in a follow-up PR). As for the operations for which we currently support driver delegation (import/export/generate/sign_hash/verify_hash/cipher) we need support for hash operations in the transparent test driver. Due to the code reorganization of this PR the transparent test driver hash functions should just call mbedtls_transparent_test_driver_hash_xyz() functions from psa_crypto_hash.c functions. Please have a look in psa_crypto_rsa/ecp.c to see how mbedtls_transparent_test_driver_hash_xyz() functions should relate to mbedtls_psa_hash_xyz() functions and in particular the (BEYOND THIS POINT ... sections). This is a trick we have to currently do some testing of the PSA configuration with test accelerators (see the test_psa_crypto_config_basic component in all.sh) while we don't support double compilation (one compilation for the library and one compilation for the test drivers) in the tests.

stevew817 commented on include/psa/crypto_struct.h:
```diff
@@ -89,35 +89,10 @@ typedef struct {
-        mbedtls_sha512_context sha512;
-#endif
-    } ctx;
+    psa_operation_driver_context_t ctx;
```
I'd like to make a case for rather applying this format and re-organising cipher.

The reasoning is stack usage. If we keep the gigantic union, it will expand to the size of the largest context supported in software, regardless of which algorithm you actually are using at that time. If you only need SHA-512 sporadically, it doesn't make sense to stack-allocate 212 bytes (on a 32-bit system) for each hash operation when the rest of them can be hardware-accelerated with potentially as little as 12 bytes.

Can you elaborate on what exactly the mbed TLS implementation of PSA Crypto cares about with regards to memory allocation for operation contexts? The way I see it, we are doing memory allocation for driver operation contexts already. When a software implementation behaves like a pseudo-driver, I don't see why it would then need special treatment for its context structures, allocation-wise. Having dynamic allocation for *all* contexts passing through the driver-wrapper layer makes it in addition more homogenous to implement.

ronald-cron-arm replied:
I understand your point. For sure there is a tradeoff here. @gilles-peskine-arm it would be great to have your opinion on this.

> Can you elaborate on what exactly the mbed TLS implementation of PSA Crypto cares about with regards to memory allocation for operation contexts?
I think there is a long term goal to try to not do any dynamic memory allocation in the Mbed TLS PSA crypto implementation for robustness/security reasons.

> When a software implementation behaves like a pseudo-driver, I don't see why it would then need special treatment for its context structures, allocation-wise. Having dynamic allocation for all contexts passing through the driver-wrapper layer makes it in addition more homogeneous to implement.

I didn't think about it thoroughly but the other way around when compiling the library for a given set of drivers, we can imagine that the driver context types are added to the operation structure union and thus there is no dynamic allocation neither for software implementation nor for drivers.

gilles-peskine-arm replied:
There is significant demand for eliminating heap allocation in the crypto part of Mbed TLS. A design goal of the PSA crypto API was to make it reasonable to implement without heap allocation. In Mbed TLS, we're far from not needing dynamic allocation in asymmetric crypto. On the other hand, for symmetric crypto, the low-level layers (`aes.c`, `sha512.c`, etc.) don't do any dynamic allocation, and PSA code should not do any dynamic allocation for individual operations outside the parts that are still using the classic `cipher` module.

I didn't look at the details here, but if `psa_hash_setup` starts calling malloc, that's a blocker. The only allocation should be the allocation of the context structure which is done by the application.

Yes, as a consequence, the size of the operation structure is the largest needed size among supported algorithms. This is usually not a big concern because constrained platforms where stack or heap size is at a premium usually don't implement algorithms with large memory consumption.

stevew817 replied:
If that's the design goal, then all previous driver work needs to be revisited as well, since all plugged-in drivers with multipart support would need to contribute their respective operation context sizes to the overarching struct union in `crypto_struct.h`. I don't see an immediate way of achieving that.

The obvious way to go about it would be to make `psa_driver_wrappers.h` be autogenerated as well, and have it contain includes for all drivers, as well as define the union of all context structures.

At this point, if it's a stated design goal to not dynamically allocate operation contexts, I'm wondering why the current driver wrapper implementation even got put in place with its dynamic allocation of driver contexts.

ronald-cron-arm replied:
For the time being in the driver wrapper there are dynamic allocations only for the driver context of cipher multi-part operations. That's ok I think for the time being to expand this to the other multi-part driver operation contexts but better to not go beyond that as that not really the way we want it to be eventually. Then, when implementing the generation of the wrapper, we can revisit that and try and remove the dynamic allocations for the multi-part driver operation contexts. I have just amended #3342 to not forget about it. Does it seem a reasonable plan?

stevew817 replied:
I don't think I'm on board with that. It's extremely wasteful to have *both* those 212 bytes always on the stack and then allocate an additional 212 bytes (for arguments' sake) on the heap when the wrapper figures out it can use a driver.

Those allocations should be aliased, either through always putting them on the heap (which you apparently don't want) or by making the context size for all compiled-in drivers visible to `crypto_struct.h`. Unless someone sees a third solution which avoids needlessly allocating large swaths of memory?

ronald-cron-arm replied:
Why an additional 212 bytes on the heap ? The driver context should be only of the size needed by the driver.

stevew817 replied:
Yes, it will only be the size needed by the driver. But like I said, for argument's sake, let's assume that the driver needs to store the same amount of context as the software implementation would, i.e. 212.

gilles-peskine-arm replied:
> by making the context size for all compiled-in drivers visible to `crypto_struct.h`. 

That's the plan.

stevew817 replied:
Okay, so if that's the plan, then all work on the driver dispatch should adhere to the plan, right? Why wasn't this an issue 6 months ago on #3644 then?

Frustration aside: how do you see this progressing? By updating this PR to declare `psa_driver_wrappers.h` as future autogenerated file, move the #includes for the different drivers up into the header file, and make `crypto_struct.h` depend on the driver headers?

ronald-cron-arm replied:
I've proposed a way forward above where we keep allocating the driver contexts while we don't have the wrapper generation. Definitely not a possible option on your side? On first sight, the two times 212 bytes of memory seem a bit of an edge case to me where you want to do the most memory expensive hash operation both in software and in hardware.

gilles-peskine-arm replied:
#3644 was about symmetric ciphers, which currently go through the `mbedtls_cipher` abstraction which calls calloc, so it wasn't a regression. This PR is about hashes, which are already malloc-free.

When the driver wrapper generation code is available, `crypto_struct.h` will include a header created by the generation code. In the meantime, there needs to be public a header file that defines the driver context types, which driver vendors will have to edit manually.

gilles-peskine-arm replied:
> I've proposed a way forward above where we keep allocating the driver contexts while we don't have the wrapper generation. Definitely not a possible option on your side?

That's fine by me as long as users who don't use drivers aren't affected. If you use the software implementation, calling `psa_hash_setup` must not start causing a heap allocation.

stevew817 replied:
> Definitely not a possible option on your side?

Not on this scale. With the cipher operation PSA context the overhead was excusable because the cipher context itself contains dynamically-allocated contexts as Gilles pointed out (though it still isn't small by any account). With hash, where having SHA-512 implemented in fallback is a common usecase, it's too much.

Back to the drawing board, then. Would somebody care to actually document this (PSA Core wants to actively get rid of dynamic allocations) somewhere? It's new to me, at least.

gilles-peskine-arm replied:
> Would somebody care to actually document this (PSA Core wants to actively get rid of dynamic allocations) somewhere?

I've started compiling a list of such things at https://github.com/ARMmbed/mbedtls-docs/pull/5

stevew817 commented on library/psa_crypto_hash.h:
```diff
@@ -0,0 +1,248 @@
+ *         Success.
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The operation state is not valid (it must be active).
+ * \retval #PSA_ERROR_BUFFER_TOO_SMALL
```
I'm not 100% convinced about that. It's this function which is responsible for the copy to the output buffer, so it would make sense that it validates the length before copy? What else is the reasoning for supplying a buffer size argument?

stevew817 commented on library/psa_crypto_hash.h:
```diff
@@ -0,0 +1,248 @@
+ * \retval #PSA_ERROR_NOT_SUPPORTED
+ *         \p alg is not supported or is not a hash algorithm.
+ * \retval #PSA_ERROR_INVALID_ARGUMENT
+ * \retval #PSA_ERROR_BUFFER_TOO_SMALL
```
The write-to-output action is happening inside this function, so I would argue it has a responsibility to not write outside of the buffer boundary.

stevew817 commented on library/psa_crypto_hash.c:
```diff
@@ -0,0 +1,442 @@
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    size_t actual_hash_length = PSA_HASH_LENGTH( operation->alg );
+
+    /* Fill the output buffer with something that isn't a valid hash
```
As in my other comments, I'd like to keep the buffer length checking as close to the function which actually does the output as possible.

I could move the memset up into psa_crypto, but that's the part you're not sure about, so...

stevew817 commented:
@ronald-cron-arm I believe I have addressed all of your feedback now.

* Dynamic allocation of driver contexts is gone, and a new paradigm for declaring driver context structures has been set up
* Test driver hooks are added for hash operations
* Superfluous operation pointer null-checks have been removed
* Input argument checking has been moved back into psa_crypto.c where it made sense (i.e. all of the ones you pointed at minus the output buffer checking, which I believe belongs with the code that is doing the actual output).

ronald-cron-arm commented:
A few jobs are failing. All the ones have checked have the same error:

psa_crypto_hash.c: In function 'mbedtls_psa_hash_update':
psa_crypto_hash.c:215:12: error: unused parameter 'input_length' [-Werror=unused-parameter]
     size_t input_length )

stevew817 commented:
Thanks, should be fixed now

ronald-cron-arm reviewed (changes requested):
Thanks for the changes. I am happy with the way things are organized now. I find psa_crypto.c and psa_crypto_driver_wrappers.c code is simple and easy to follow. There is still a bit of pre-compiler guards tweaks to do in psa_crypto_hash.c due to the way we currently test PSA config and driver accelerators. I've tried to explain what is missing and you have psa_crypto_rsa/ecp.c/h as examples but feel free to ask if you have questions. Then, the last thing will be to update the all.sh test_psa_crypto_config_basic and test_psa_crypto_drivers components with full hash acceleration (see https://github.com/ARMmbed/mbedtls/pull/4151/commits/17b3afcc336e94365b3fa9dc5b4ab7e0af6b8591 as an example).

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -1082,4 +1085,174 @@ psa_status_t psa_driver_wrapper_cipher_abort(
+    size_t hash_size,
+    size_t *hash_length)
+{
+    psa_status_t status = PSA_ERROR_NOT_SUPPORTED;
```
No. The setup here is to try each accelerator which has a `hash_compute` entry point in turn. If there are no drivers with such entry point, then `PSA_ERROR_NOT_SUPPORTED` is correct.

stevew817 replied:
OK in combination with the other change.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -1082,4 +1085,174 @@ psa_status_t psa_driver_wrapper_cipher_abort(
+
+        return( PSA_ERROR_NOT_SUPPORTED );
+    }
+    return( status );
```
Well, OK then, if that's how you want it.

stevew817 commented on library/psa_crypto_hash.c:
```diff
@@ -0,0 +1,616 @@
+    if( hash_size != 0 )
+        memset( hash, '!', hash_size );
+
+    if( hash_size < actual_hash_length )
```
The core doesn't have that information anymore, since the algorithm in use is stored by the driver context, and not the overarching context.

gilles-peskine-arm commented on include/psa/crypto_struct.h:
```diff
@@ -68,14 +68,9 @@ extern "C" {
-#include "mbedtls/sha512.h"
+
+/* Include the context definition for the compiled-in drivers */
+#include "../../library/psa_crypto_driver_wrappers_contexts.h"
```
To use Mbed TLS, it is sufficient to have the headers in `include` and compiled libraries (e.g. `libmbed*.a` or `libmbed*.so` `libmbed*.dll`). In particular, headers in `include` may not depend on headers anywhere else in the source tree.

gilles-peskine-arm replied:
`psa_crypto_driver_wrappers_contexts` needs to move to `include/psa`, yes. But it also needs to lose some things, including references to the test drivers. And yes, drivers need to place their context information under `include`. The driver wrapper generator will produce both `include/psa/something.h` and `library/something.c` as outputs.

stevew817 replied:
Test driver contexts are driver contexts, too. So how would `psa_crypto_driver_wrappers_contexts` move to `include/psa` and simultaneously drop the test driver references?

stevew817 replied:
Not doing dynamic allocation means all driver contexts need to be fully-typed by the time the driver context union in the struct gets evaluated. So all driver contexts (including the test driver's) need to be reachable from (and thus according to your stated goal, located inside of) the library's `/include` folder.

So... Are we back to creating a `/include/psa/drivers/` folder where all drivers (including both mbedTLS software-based and test) drivers need to put their context structure declarations? And the autogenerated `include/psa/psa_crypto_driver_wrappers_contexts.h` will pull in all declared drivers' context structure headers?

gilles-peskine-arm replied:
Namespacing drivers as `psa/drivers/*.h` seems sensible to me. Mbed TLS itself wouldn't include a `include/psa/drivers` directory though, since we'd have nothing to put there. (Unless we put headers for the built-in implementations there?) If you package Mbed TLS with your own drivers, is it up to Mbed TLS to tell you where to put your headers?

> And the autogenerated include/psa/psa_crypto_driver_wrappers_contexts.h will pull in all declared drivers' context structure headers?

Yes.

stevew817 replied:
> Unless we put headers for the built-in implementations there

That's exactly what I was saying. They'd need to go there, since apparently all of their structures need to be declared in `/include`, and just dumping them flat in `include/psa` will (IMO) be terribly messy.

gilles-peskine-arm reviewed (changes requested):
There's a fundamental problem with the way headers are structured. All declarations and definitions needed to compile a program using the library must be in `include/**/*.h`.

ronald-cron-arm commented:
CI is OK no genuine failure.

stevew817 commented:
Does this fit the bill now, @gilles-peskine-arm ?

stevew817 commented:
Rebased due to merge conflict in mbedTLS.vcxproj. State before rebase can be found here: https://github.com/stevew817/mbedtls/tree/dispatch_hash_operation_rebase_1

Rebase was clean, i.e. none of my commits changed content.

gilles-peskine-arm commented on include/CMakeLists.txt:
```diff
@@ -13,6 +14,10 @@ if(INSTALL_MBEDTLS_HEADERS)
         PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ)
 
+    install(FILES ${psa_driver_headers}
+        DESTINATION include/psa/drivers
```
The change in installation instructions needs to be documented in the changelog. This affects all applications that use any part of PSA crypto, even when no (non-builtin) drivers are present.

stevew817 replied:
Backed out the change introducing the new folder. It's just going to be delaying the driver work and can be done another time.

gilles-peskine-arm commented on include/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@ if(INSTALL_MBEDTLS_HEADERS)
 
     file(GLOB headers "mbedtls/*.h")
     file(GLOB psa_headers "psa/*.h")
+    file(GLOB psa_driver_headers "psa/drivers/*.h")
```
The `psa` header file namespace is shared with all the PSA components. So its use is not solely a decision for PSA crypto.

This pull requests only introduces files matching `psa/drivers/crypto*`, which is enough namespacing to avoid conflicts with other PSA components. But what names will external drivers use? Should each external driver get its own subdirectory? Should external driver header names have a mandatory prefix like `crypto_driver_*`?

stevew817 replied:
Backed out the change introducing the new folder. It's just going to be delaying the driver work and can be done another time.

gilles-peskine-arm reviewed (dismissed):
This pull request performs a significant architectural change: it adds a new directory of header files. This affects users with custom build, installation or packaging code. This also creates a potential namespace conflicts with other PSA projects. Before doing such a significant change, we need to think about the implications and the interactions with other projects. Given the impact on users, we shouldn't introduce a new directory only to rename or remove it later.

I recommend splitting this work in two parts: one that does the refactoring necessary to better support drivers with no impact on applications that don't use drivers, and a future one to handle driver header installation. The first part should not introduce any new header directory. Introducing new header files that aren't directly used by applications in `include/psa` is fine, and we can rename them later.

stevew817 commented on include/CMakeLists.txt:
```diff
@@ -13,6 +14,10 @@ if(INSTALL_MBEDTLS_HEADERS)
         PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ)
 
+    install(FILES ${psa_driver_headers}
+        DESTINATION include/psa/drivers
```
Backed out the change introducing the new folder. It's just going to be delaying the driver work and can be done another time.

gilles-peskine-arm reviewed (commented):
With no header directory being created, I'm now happy with the design. I haven't done a full code review.

stevew817 commented:
Rebased due to merge conflict in `all.sh`. That's the only thing I fixed, all other commit content remained untouched.

stevew817 commented:
@ronald-cron-arm @tom-daubney-arm @mstarzyk-mobica Given the PR now has design approval, it's ready for your (re-)review.

ronald-cron-arm reviewed (changes requested):
Here is my second round of review following the rework around the PSA hash operation context. The main comment is the one in config_psa.h.

tom-daubney-arm reviewed (dismissed):
I am approving this PR subject to work being undertaken to satisfactorily address the issues that arose from @ronald-cron-arm's previous review.
I have nothing further to add beyond those comments and as such approve this PR to reflect this. I will re-review once these issues have been addressed. Thanks.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -1066,4 +1072,171 @@ psa_status_t psa_driver_wrapper_cipher_abort(
+
+    /* Try setup on accelerators first */
+#if defined(PSA_CRYPTO_DRIVER_TEST)
+    status = mbedtls_transparent_test_driver_hash_setup( &operation->ctx.test_ctx, alg );
```
I'll fix it, but this is supposedly an autogenerated file. Considering there's no delinter for mbed TLS yet, is the line length going to be enforced at autogeneration?

gilles-peskine-arm replied:
In my (not nearly working) draft for autogeneration, I make no provision to break lines inside an instruction. So no, there won't be a maximum line length.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -1066,4 +1072,171 @@ psa_status_t psa_driver_wrapper_cipher_abort(
+    if( status == PSA_SUCCESS )
+        operation->id = PSA_CRYPTO_MBED_TLS_DRIVER_ID;
+
+    if( status != PSA_ERROR_NOT_SUPPORTED )
```
Then some compilers will complain about an unreachable statement (the voiding and return `NOT_SUPPORTED` under here). Probably fixable by putting that in a `#else` block, but I'm not sure it'll be that pretty.

stevew817 replied:
Trying to do this with preprocessor logic isn't pretty at all, so I'd rather keep it as-is.

ronald-cron-arm replied:
Indeed.

stevew817 commented:
All of @ronald-cron-arm 's feedback should now be addressed. I rebased as per the rewrite request, and added the other changes as new commits.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -2141,363 +2142,67 @@ psa_status_t psa_copy_key( mbedtls_svc_key_id_t source_key,
+    if( !PSA_ALG_IS_HASH( alg ) )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    /* Ensure all of the context is zeroized, not just the dummy int */
```
Not sure what you mean by "not just the dummy int".

stevew817 replied:
When a user initialises the context structure with `PSA_HASH_OPERATION_INIT`, the macro is built in such a way that the underlying union `psa_driver_hash_context_t` can be initialised with a single item, to avoid the init macro from having to know the details of the other union member's content.

But this also means the only thing that's guaranteed to be set to zero in that union is the unsigned `dummy` member of the union, and not the entire struct, therefore the need for a full memset (to guarantee that the driver code gets a fully-zeroized context structure upon a `setup` call).

stevew817 replied:
Updated the language to be clearer.

ronald-cron-arm commented on library/psa_crypto_hash.c:
```diff
@@ -0,0 +1,712 @@
+ * the core implementation should be present. */
+#if defined(MBEDTLS_PSA_BUILTIN_HASH) || \
+    defined(INCLUDE_HASH_TEST_DRIVER)
+#define INCLUDE_HASH_CORE       1
```
Can be removed (used only once) and replaced by `#if defined(MBEDTLS_PSA_BUILTIN_HASH) || defined(PSA_CRYPTO_DRIVER_TEST) below.

ronald-cron-arm commented on library/psa_crypto_hash.c:
```diff
@@ -0,0 +1,712 @@
+ * entry points need to be implemented.  */
+#if defined(PSA_CRYPTO_DRIVER_TEST) && \
+    defined(MBEDTLS_PSA_ACCEL_HASH)
+#define INCLUDE_HASH_TEST_DRIVER
```
I think we can simplify here, just use PSA_CRYPTO_DRIVER_TEST to guard the test driver entry points and don't define INCLUDE_HASH_TEST_DRIVER and MBEDTLS_PSA_ACCEL_HASH. We don't really care to remove the entry points if there is no need for test driver hash acceleration and "is_hash_accelerated" makes sure that we will return in error if an hash operation it is not supposed to support is asked to the test driver.

stevew817 replied:
'just' `PSA_CRYPTO_DRIVER_TEST` is not possible, since you'd get an undefined reference on the hash core if compiling without any hash algorithms turned on (e.g. an AES-only build). So it invariably needs to be `PSA_CRYPTO_DRIVER_TEST && MBEDTLS_PSA_ACCEL_HASH`

ronald-cron-arm commented on include/psa/crypto_builtin_hash.h:
```diff
@@ -0,0 +1,101 @@
+    defined(MBEDTLS_PSA_ACCEL_ALG_SHA_256) || \
+    defined(MBEDTLS_PSA_ACCEL_ALG_SHA_384) || \
+    defined(MBEDTLS_PSA_ACCEL_ALG_SHA_512)
+#define MBEDTLS_PSA_ACCEL_HASH
```
Better to remove this I think, see comments in psa_crypto_hash.c.

ronald-cron-arm commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -1066,4 +1071,172 @@ psa_status_t psa_driver_wrapper_cipher_abort(
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    /* Try accelerators first */
+#if defined(PSA_CRYPTO_DRIVER_TEST) && defined(MBEDTLS_PSA_ACCEL_HASH)
```
Eventually, just remove MBEDTLS_PSA_ACCEL_HASH I think, see comments in psa_crypto_hash.c.

stevew817 replied:
Which is either going to cause an undefined reference for the test driver entry point (if the entry points keep being guarded with DRIVER_TEST && ACCEL_HASH), or internally inside psa_crypto_hash.c. Unless you are trying to say that the hash core needs to be present on PSA_CRYPTO_DRIVER_TEST? Isn't that overreaching a bit?

ronald-cron-arm reviewed (changes requested):
Thanks for the changes. The CI is ok: it is expected for ABI-API-checking to fail and the other failure is an infrastructure glitch. Almost good to me, I would just prefer to simplify the guarding of the test driver hash entry points.

stevew817 commented:
@ronald-cron-arm I updated the define guards like you asked.

ronald-cron-arm reviewed (approved):
Thanks for all your work on this. LGTM.

ronald-cron-arm commented:
@tom-daubney-arm Please have a look to the last version.

tom-daubney-arm reviewed (approved):
Have followed through all the new changes and now all LGTM. Thanks.

ronald-cron-arm commented:
@mpg @gilles-peskine-arm @yanesca Could you have the final look to this PR and merge it if this is ok ?

gilles-peskine-arm commented on include/psa/crypto_builtin_hash.h:
```diff
@@ -0,0 +1,89 @@
+/*
+ *  Context structure declaration of the software-based driver which performs
+ *  hashing through the PSA Crypto driver dispatch layer.
+ */
```
Please state that this header and its content are not part of the Mbed TLS API and applications must not depend on it.

gilles-peskine-arm commented on include/psa/crypto_driver_contexts.h:
```diff
@@ -0,0 +1,51 @@
+ *  interface.
+ *
+ *  Warning: This file will be auto-generated in the future.
+ */
```
Please state that this header and its content are not part of the Mbed TLS API and applications must not depend on it.

gilles-peskine-arm commented on include/psa/crypto_builtin_hash.h:
```diff
@@ -0,0 +1,89 @@
+/*
+ *  Context structure declaration of the software-based driver which performs
+ *  hashing through the PSA Crypto driver dispatch layer.
```
Why a separate header file for each type of algorithm? This seems needlessly complex to me, and not necessarily generalizable to other types of algorithms (e.g. RSA-related definitions may be duplicated between asymmetric encryption and signature, ECC-related definitions between key agreement and signature, block cipher between cipher, AEAD and MAC).

gilles-peskine-arm commented on library/psa_crypto_hash.c:
```diff
@@ -0,0 +1,697 @@
+  */
+#if defined(PSA_CRYPTO_DRIVER_TEST)
+
+psa_status_t is_hash_accelerated( psa_algorithm_t alg )
```
This should be `static` since it isn't used in other files. In fact I'm surprised the CI let it through: I thought we ran some clang builds in a picky mode where it requires a prototype for non-static functions.

gilles-peskine-arm reviewed (commented):
I have done a design review (basically, I read the header files and looked at what functions are being created and moved around, without checking all details). Consider this approved for design.

If this was a full review with no urgency, I'd leave some minor change requests. Because none of these changes are actual bugs or raise commitments to backward compatibility, I'm going to go ahead and merge this pull request, but I've filed an issue for doing these minor changes: https://github.com/ARMmbed/mbedtls/issues/4242.
